rmw_connextdds_common/src/common/rmw_impl.cpp

```diff
@@ -1366,52 +1413,54 @@ RMW_Connext_Subscriber::finalize()
     }
   }
 
-  if (DDS_RETCODE_OK !=
+  if (nullptr != this->dds_reader && DDS_RETCODE_OK !=
     DDS_Subscriber_delete_datareader(
       this->dds_subscriber(), this->dds_reader))
   {
     RMW_CONNEXT_LOG_ERROR_SET("failed to delete DDS DataReader")
     return RMW_RET_ERROR;
   }
 
-  DDS_DomainParticipant * const participant = this->dds_participant();
+  if (!reset_cft) {
+    DDS_DomainParticipant * const participant = this->dds_participant();
 
-  if (nullptr != this->dds_topic_cft) {
-    rmw_ret_t cft_rc = rmw_connextdds_delete_contentfilteredtopic(
-      ctx, participant, this->dds_topic_cft);
+    if (nullptr != this->dds_topic_cft) {
+      rmw_ret_t cft_rc = rmw_connextdds_delete_contentfilteredtopic(
+        ctx, participant, this->dds_topic_cft);
 
-    if (RMW_RET_OK != cft_rc) {
-      return cft_rc;
+      if (RMW_RET_OK != cft_rc) {
+        return cft_rc;
+      }
+      this->dds_topic_cft = nullptr;
     }
-  }
 
-  if (this->created_topic) {
-    DDS_Topic * const topic = this->dds_topic;
+    if (this->created_topic) {
+      DDS_Topic * const topic = this->dds_topic;
 
-    RMW_CONNEXT_LOG_DEBUG_A(
-      "deleting topic: name=%s",
-      DDS_TopicDescription_get_name(
-        DDS_Topic_as_topicdescription(topic)))
+      RMW_CONNEXT_LOG_DEBUG_A(
+        "deleting topic: name=%s",
+        DDS_TopicDescription_get_name(
+          DDS_Topic_as_topicdescription(topic)))
 
-    DDS_ReturnCode_t rc =
-      DDS_DomainParticipant_delete_topic(participant, topic);
+      DDS_ReturnCode_t rc =
+        DDS_DomainParticipant_delete_topic(participant, topic);
 
-    if (DDS_RETCODE_OK != rc) {
-      RMW_CONNEXT_LOG_ERROR_SET("failed to delete DDS Topic")
-      return RMW_RET_ERROR;
+      if (DDS_RETCODE_OK != rc) {
+        RMW_CONNEXT_LOG_ERROR_SET("failed to delete DDS Topic")
+        return RMW_RET_ERROR;
+      }
     }
-  }
-
-  rmw_ret_t rc = RMW_Connext_MessageTypeSupport::unregister_type_support(
-    this->ctx, participant, this->type_support->type_name());
 
-  if (RMW_RET_OK != rc) {
-    return rc;
-  }
+    rmw_ret_t rc = RMW_Connext_MessageTypeSupport::unregister_type_support(
+      this->ctx, participant, this->type_support->type_name());
 
-  delete this->type_support;
-  this->type_support = nullptr;
+    if (RMW_RET_OK != rc) {
+      return rc;
+    }
 
+    delete this->type_support;
+    this->type_support = nullptr;
```

Comment:
The type support shouldn't be deleted if `reset_cft == true`. Extend the block above to include these lines.

Reply:
the `type_support` block is already extented into the `if(!reset_cft)` branch

Reply:
sorry, my bad